diff --git a/data/zsh-completion.in b/data/zsh-completion.in
index fe621cb..bbe9d07 100644
--- a/data/zsh-completion.in
+++ b/data/zsh-completion.in
@@ -26,21 +26,21 @@ _arguments -C -s -S "$all_opts[@]" && ret=0
 
 local PLUGIN
 local -a exts
-for PLUGIN in @PLUGINDIR@/lib*.so; do
+for PLUGIN in @PLUGINDIR@/lib*.dylib; do
   case ${PLUGIN##*/} in
-    libpdf-poppler.so)
+    libpdf-poppler.dylib)
       exts+=( pdf PDF )
       ;;
-    libpdf-mupdf.so)
+    libpdf-mupdf.dylib)
       exts+=( pdf PDF epub oxps )
       ;;
-    libps.so)
+    libps.dylib)
       exts+=( ps eps epsi epsf )
       ;;
-    libdjvu.so)
+    libdjvu.dylib)
       exts+=( djvu djv )
       ;;
-    libcb.so)
+    libcb.dylib)
       exts+=( cb7 cbr cbz cbt rar zip 7z tar )
       ;;
   esac
diff --git a/meson.build b/meson.build
index 9d52fc8..fcf8e22 100644
--- a/meson.build
+++ b/meson.build
@@ -74,9 +74,9 @@ defines = [
   '-D_DEFAULT_SOURCE',
 ]
 
-if host_machine.system() == 'darwin'
-  defines += '-DGTKOSXAPPLICATION'
-endif
+# if host_machine.system() == 'darwin'
+#   defines += '-DGTKOSXAPPLICATION'
+# endif
 
 # compile flags
 flags = [
@@ -168,6 +168,15 @@ headers = files(
 )
 headers += version_header
 
+if host_machine.system() == 'darwin'
+  sources += files(
+    'zathura/mac-integration.c',
+    )
+  headers += files(
+    'zathura/mac-integration.h',
+    )
+endif
+
 # zathura helper library
 libzathura = static_library('zathura',
   sources + ['zathura/database-sqlite.c'],
diff --git a/zathura/main.c b/zathura/main.c
index a439a78..fc4dc23 100644
--- a/zathura/main.c
+++ b/zathura/main.c
@@ -1,7 +1,10 @@
 /* SPDX-License-Identifier: Zlib */
 
 #ifdef GTKOSXAPPLICATION
-#include <gtkosxapplication.h>
+// #include <gtkosxapplication.h>
+#endif
+#ifdef MAC_INTEGRATION
+#include "mac-integration.h"
 #endif
 
 #include <girara-gtk/settings.h>
@@ -284,17 +287,21 @@ GIRARA_VISIBLE int main(int argc, char* argv[]) {
   }
 
 #ifdef GTKOSXAPPLICATION
-  GtkosxApplication* zathuraApp = g_object_new(GTKOSX_TYPE_APPLICATION, NULL);
-  gtkosx_application_set_use_quartz_accelerators(zathuraApp, FALSE);
-  gtkosx_application_ready(zathuraApp);
-  {
-    const gchar* id = gtkosx_application_get_bundle_id();
-    if (id != NULL) {
-      girara_warning("TestIntegration Error! Bundle has ID %s", id);
-    }
-  }
+//   GtkosxApplication* zathuraApp = g_object_new(GTKOSX_TYPE_APPLICATION, NULL);
+//   gtkosx_application_set_use_quartz_accelerators(zathuraApp, FALSE);
+//   gtkosx_application_ready(zathuraApp);
+//   {
+//     const gchar* id = gtkosx_application_get_bundle_id();
+//     if (id != NULL) {
+//       girara_warning("TestIntegration Error! Bundle has ID %s", id);
+//     }
+//   }
 #endif // GTKOSXAPPLICATION
 
+#ifdef MAC_INTEGRATION
+  zathura_mac_init(zathura);
+#endif
+
   /* run zathura */
   gtk_main();
   return 0;
